src/persist/src/indexed/mod.rs

```diff
@@ -205,13 +207,37 @@ impl<K: Data, V: Data, U: Buffer, L: Blob> Indexed<K, V, U, L> {
         })?;
 
         for (id, updates) in updates_by_id.drain() {
+            let future = self
+                .futures
+                .get_mut(&id)
+                .ok_or_else(|| Error::from(format!("never registered: {:?}", id)))?;
+
+            // We maintain the invariant that futures_seqno_upper is >= every
+            // future's seqno_upper and that there is nothing for that future in
+            // [future.seqno_upper, self.futures_seqno_upper). Use this to make the
+            // seqnos of all the future batches line up.
+            debug_assert_eq!(desc.start, self.futures_seqno_upper);
+            let new_start = future.seqno_upper()[0];
+            debug_assert!(new_start <= desc.start);
+            let mut desc = desc.clone();
+            desc.start = new_start;
```

Comment:
I think you could use the `since` field to express "which subrange of sequence numbers do we actually have data for" with e.g.

```
desc.since = self.futures_seqno_upper;
```

not sure if this is useful for anything (I can't think of anything because consolidation means there may not even be any updates to place into future) but wanted to mention in case you found it useful

Reply:
the only useful property i can think of with sequence number descriptions is:

given a range of sequence numbers say [lo, hi)

it would be nice to verify that all of the sequence numbers in that range are covered by exactly one future batch (aka no sequence number's update was orphaned and no sequence number's update was read by more than 1 stream)

I think to do this we would have to allow empty future batches (for cases when inputs consolidated to the empty sets) and gaps between future batches (to ensure that exactly one future "claimed" a sequence number range).

thoughts?

Reply:
> desc.since = self.futures_seqno_upper;

Okay good nerd snipe. I just went off and read dd code for a while. My understanding here is still shaky but given that the logic for merging two dd trace batches is to `join` the since fields, I think this is not a correct usage of since. Maybe?

> given a range of sequence numbers say [lo, hi) it would be nice to verify that all of the sequence numbers in that range are covered by exactly one future batch (aka no sequence number's update was orphaned/left unread and no sequence number's update was read by more than 1 stream)

Yup I agree with this.

> I think to do this we would have to allow empty future batches (for cases when inputs consolidated to the empty sets)

Yeah okay, so that's another one of our sus invariants to bite the dust.

> gaps between future batches (to ensure that exactly one future "claimed" a sequence number range)

Hmm? I don't follow this part

Reply:
So what I'm trying to say is -- imagine we have two streams 1 and 2

lets say sequence numbers [0,3) are inserts into stream 1, [3, 4) is an insert into stream 2, [4, 5) is an insert into stream 1, and then [5, 7) are inserts into stream 2

as we construct future batches we want to be able to enforce "every sequence number was read by exactly one stream" which means

stream 1 gets future batches [0, 3) and [4,5)
stream 2 gets future batches at [3, 4) and [5, 7)

now the future batches in each individual stream cover increasing and non-overlapping ranges of sequence numbers but they have gaps which we don't currently allow. But I think this is still useful because now we can look at the future batches across all streams to say "all updates between sequence numbers [0, 7) have been read into exactly one stream's future". If we instead forced each stream's future to have contiguous batches then we would lose the ability to recover that check, and I'm not entirely sure what we gain in return.

Reply:
In case its more clear with code I wrote up a quick patch to describe what I'm trying to say:

```
         // Overlapping
altaria-2:persist Test$ git diff
diff --git a/src/persist/src/indexed/encoding.rs b/src/persist/src/indexed/encoding.rs
index 5e4756bf9..3c41089ae 100644
--- a/src/persist/src/indexed/encoding.rs
+++ b/src/persist/src/indexed/encoding.rs
@@ -226,6 +226,35 @@ impl BlobMeta {
                 )));
             }
         }
+
+        let mut future_descs = vec![];
+        for (_, meta) in futures.iter() {
+            for (desc, _) in meta.batches.iter() {
+                let lower = desc.lower()[0];
+                let upper = desc.upper()[0];
+
+                // Inserting tuples because we can't sort
+                // Descriptions
+                future_descs.push((lower, upper));
+            }
+        }
+
+        future_descs.sort();
+        let mut prev: Option<(SeqNo, SeqNo)> = None;
+
+        for desc in future_descs {
+            if let Some(prev) = prev {
+                if prev.1 != desc.0 {
+                    return Err(Error::from(format!(
+                        "invalid future desc sequence {:?} {:?}",
+                        prev, desc
+                    )));
+                }
+            }
+
+            prev = Some(desc);
+        }
+
         Ok(())
     }
 }
@@ -246,7 +275,7 @@ impl BlobFutureMeta {
         let mut prev: Option<&Description<SeqNo>> = None;
         for (desc, _) in self.batches.iter() {
             if let Some(prev) = prev {
-                if prev.upper() != desc.lower() {
+                if !PartialOrder::less_equal(prev.upper(), desc.lower()) {
                     return Err(format!(
                         "invalid batch sequence: {:?} followed by {:?}",
                         prev, desc
@@ -312,11 +341,6 @@ where
         if PartialOrder::less_equal(self.desc.upper(), &self.desc.lower()) {
             return Err(format!("invalid desc: {:?}", &self.desc).into());
         }
-        // TODO: It's unclear if this invariant is useful/harmful. Feel free to
-        // remove it if it ends up not making sense.
-        if self.updates.is_empty() {
-            return Err("updates is empty".into());
-        }

         let mut prev: Option<(&u64, &K, &V)> = None;
         for update in self.updates.iter() {
@@ -457,14 +481,6 @@ mod tests {
         };
         assert_eq!(b.validate(), Ok(()));

-        // Empty
-        let b: BlobFutureBatch<String, String> = BlobFutureBatch {
-            id: Id(0),
-            desc: seqno_desc(0, 2),
-            updates: vec![],
-        };
-        assert_eq!(b.validate(), Err("updates is empty".into()));
-
         // Invalid desc
         let b: BlobFutureBatch<String, String> = BlobFutureBatch {
             id: Id(0),
@@ -669,12 +685,7 @@ mod tests {
             ts_lower: Antichain::from_elem(0),
             batches: vec![(seqno_desc(0, 1), "".into()), (seqno_desc(2, 3), "".into())],
         };
-        assert_eq!(
-            b.validate(),
-            Err(Error::from(
-                "invalid batch sequence: Description { lower: Antichain { elements: [SeqNo(0)] }, upper: Antichain { elements: [SeqNo(1)] }, since: Antichain { elements: [SeqNo(0)] } } followed by Description { lower: Antichain { elements: [SeqNo(2)] }, upper: Antichain { elements: [SeqNo(3)] }, since: Antichain { elements: [SeqNo(0)] } }"
-            ))
-        );
+        assert_eq!(b.validate(), Ok(()),);

         // Overlapping
         let b = BlobFutureMeta {
diff --git a/src/persist/src/indexed/mod.rs b/src/persist/src/indexed/mod.rs
index 5b087f5ed..f174f6d6d 100644
--- a/src/persist/src/indexed/mod.rs
+++ b/src/persist/src/indexed/mod.rs
@@ -237,10 +237,6 @@ impl<K: Data, V: Data, U: Buffer, L: Blob> Indexed<K, V, U, L> {
         // Future batches are required to be sorted and consolidated by ((ts, (k, v)).
         differential_dataflow::consolidation::consolidate_updates(&mut updates);

-        if updates.is_empty() {
-            return Ok(());
-        }
-
         // Reshape updates back to the desired type.
         let updates: Vec<_> = updates
             .drain(..)
altaria-2:persist Test$
```

Reply:
Isn't this going to create a ton of batches? At the moment, when we call step, we create one future batch for each stream written to since the last step call. Couldn't your proposal potentially create one batch per SeqNo written since the last step call?

We could fix that particular problem with a variant of what you're saying where we require each individual stream's batches to be non-overlapping and that each SeqNo in the range is covered by at least one stream... but I have no idea how to reason about what the output description of a compactions over two non-adjacent batches would be.

I also worry that we're getting ahead of ourselves here. I'm not convinced we really understand all the requirements yet, so spending too much time on this might be premature. I vote we merge what's here, which seems to work, and revisit later. Thoughts?

Reply:
> Isn't this going to create a ton of batches? At the moment, when we call step, we create one future batch for each stream written to since the last step call. Couldn't your proposal potentially create one batch per SeqNo written since the last step call?

Yes if the writes to N streams are interleaved then my proposal would create > N future batches where the current approach makes N future batches

 > I also worry that we're getting ahead of ourselves here. I'm not convinced we really understand all the requirements yet, so spending too much time on this might be premature. I vote we merge what's here, which seems to work, and revisit later. Thoughts?

Works for me!

Reply:
> Yes if the writes to N streams are interleaved then my proposal would create > N future batches where the current approach makes N future batches

👍 I'd like to avoid that if possible :)